src/Aspire.Hosting/Pipelines/DistributedApplicationPipeline.cs

```diff
@@ -201,126 +172,165 @@ private static void ValidateSteps(IEnumerable<PipelineStep> steps)
-    /// <param name="stepsByName">A dictionary mapping step names to their corresponding step objects</param>
-    /// <returns>A list of lists where each list contains the steps to be executed at the same level</returns>
-    private static List<List<PipelineStep>> ResolveDependencies(
-        IEnumerable<PipelineStep> steps,
-        Dictionary<string, PipelineStep> stepsByName)
+    private static async Task ExecuteStepsAsTaskDag(
+        List<PipelineStep> steps,
+        Dictionary<string, PipelineStep> stepsByName,
+        DeployingContext context)
     {
-        // Initial a graph that represents a step and its dependencies
-        // and an inDegree map to count the number of dependencies that
-        // each step has.
-        var graph = new Dictionary<string, List<string>>();
-        var inDegree = new Dictionary<string, int>();
+        // Validate no cycles exist in the dependency graph
+        ValidateDependencyGraph(steps, stepsByName);
 
+        // Create a TaskCompletionSource for each step
+        var stepCompletions = new Dictionary<string, TaskCompletionSource>(steps.Count);
         foreach (var step in steps)
         {
-            graph[step.Name] = [];
-            inDegree[step.Name] = 0;
+            stepCompletions[step.Name] = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
         }
 
-        // Process all the `RequiredBy` relationships in the graph and adds
-        // the each `RequiredBy` step to the DependsOn list of the step that requires it.
-        foreach (var step in steps)
+        // Execute a step after its dependencies complete
+        async Task ExecuteStepWithDependencies(PipelineStep step)
         {
-            foreach (var requiredByStep in step.RequiredBySteps)
+            var stepTcs = stepCompletions[step.Name];
+
+            // Wait for all dependencies to complete (will throw if any dependency failed)
+            if (step.DependsOnSteps.Count > 0)
             {
-                if (!graph.ContainsKey(requiredByStep))
+                try
                 {
-                    throw new InvalidOperationException(
-                        $"Step '{step.Name}' is required by unknown step '{requiredByStep}'");
+                    var depTasks = step.DependsOnSteps.Select(depName => stepCompletions[depName].Task);
+                    await Task.WhenAll(depTasks).ConfigureAwait(false);
                 }
-
-                if (stepsByName.TryGetValue(requiredByStep, out var requiredByStepObj) &&
-                    !requiredByStepObj.DependsOnSteps.Contains(step.Name))
+                catch (Exception ex)
                 {
-                    requiredByStepObj.DependsOnSteps.Add(step.Name);
+                    // Dependency failed - mark this step as failed and stop, but don't re-throw
+                    // to avoid counting the same root cause exception multiple times
+                    stepTcs.TrySetException(ex);
```

Comment:
Maybe figure out what dependency step failed, get its name, and wrap exception? It would be a lot clearer to see the name of the dependency that failed that caused this step to fail.